cloud/amazon/iam.py

```diff
@@ -0,0 +1,354 @@
+
+    ec2_url, aws_access_key, aws_secret_key, region = get_ec2_creds(module)
+
+    try:
+        iam = boto.iam.connection.IAMConnection(
+                aws_access_key_id=aws_access_key, 
+                aws_secret_access_key=aws_secret_key,
+            )
+    except boto.exception.NoAuthHandlerFound, e:
+        module.fail_json(msg=str(e))
+
+    result = {}
+    changed = False        
+
+    group_list = []
+    for obj in iam.get_all_groups().list_groups_result.groups:
+        group_list.append(obj.group_name)
+
+    user_list = []
+    for obj in iam.get_all_users().list_users_result.users:
+        user_list.append(obj.user_name)
+
+    role_list = []
+    for obj in iam.list_roles().list_roles_response.list_roles_result.roles:
+        role_list.append(obj.role_name)
+
+##  Create, Update or Delete Users
+    if iam_type == 'user' and state == 'present':
+        if name not in user_list:
+            result = create_user(iam, name, password, path, key_state)
+            if groups is not None:
+                result['group'] = add_user_to_group(iam, name, group)
+        elif name in user_list:
+            result['user_info'] = iam.get_user(name).get_user_result.user
+            result['user_info']['access_keys'] = iam.get_all_access_keys(name).list_access_keys_result.access_key_metadata
+    elif iam_type == 'user' and state == 'absent':
+        if name in user_list:
+            delete_all_keys(iam, name)
+            result = delete_user(iam, name)
+    elif iam_type == 'user' and state == 'update':
+        if name in user_list:
+            result = update_user(iam, name, new_name, path, new_path, key_state)
+        if group is not None:
+            result['group'] = add_user_to_group(iam, name, group)
+        ## No explicit option for create defined here. It is defined in the update_user function above
+        if key_state == 'deactivate':
+            act_deact_all_keys(iam, 'inactive', name)
+        elif key_state == 'activate':
+            act_deact_all_keys(iam, 'active', name)
+
+##  Create, Update or Delete Groups
+    if iam_type == 'group' and state == 'present':
+        if name not in group_list:
+            result = create_group(iam, name, path)
+    elif iam_type == 'group' and state == 'absent':
+        if name in group_list:
+            result = delete_group(iam, name)
+    elif iam_type == 'group' and state == 'update':
+        if name in group_list:
+            result = update_group(iam, name, new_name, new_path)
+
+## Create, or Delete Roles. There is no update as boto does not currently provide a role_update function
+    if iam_type == 'role' and state == 'present':
+        if name not in role_list:
+            result = create_role(iam, name, path)
+    elif iam_type == 'role' and state == 'absent':
+        if name in role_list:
+            result = delete_role(iam, name)
+
+    module.exit_json(**result)
```

Comment:
To increase readability and maintainability of `main()`, could factor out the `iam_type`-specific code into dedicated functions. E.g., `work_on_role()`, `work_on_group()`, and `work_on_user()`.